[BugFix] Getting analyzing error when hit the expression rewrite for generated column (#29247)

srlch opened the pull request:
Problem:
If the expression in query hits the rewriting to generated column, it will get analyzing error if using a alias for the table. Because the default table name for slotRef in generated column expression will be the original one if the column is add by schema change rather than create table stmt.

Solution:
keep the expression unanalyze when schema change.

Fixes #29247

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 5 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|0|4|00.00%|[587, 589, 590, 591]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 5 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|0|4|00.00%|[587, 588, 589, 590]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 5 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|0|4|00.00%|[587, 588, 589, 590]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 5 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|0|4|00.00%|[587, 588, 589, 590]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 5 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|0|4|00.00%|[587, 588, 589, 590]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 5 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|0|4|00.00%|[587, 588, 589, 590]|

srlch committed:
[BugFix] Getting analyzing error when hit the expression rewrite for generated column (#29247)

Problem:
If the expression in query hits the rewriting to generated column, it will get analyzing error
if using a alias for the table. Because the default table name for slotRef in generated column
expression will be the original one if the column is add by schema change rather than create table stmt.

Solution:
keep the expression unanalyze when schema change.

Signed-off-by: srlch <linzichao@starrocks.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 5 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|0|4|00.00%|[587, 588, 589, 590]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 5 / 5 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnDef.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|4|4|100.00%|[]|

wanpengfei-git commented:
@Mergifyio backport branch-3.1
